java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/memory/azurecognitivesearch/AzureCognitiveSearchMemory.java

```diff
@@ -0,0 +1,355 @@
+
+        SemanticSettings semanticSettings =
+                new SemanticSettings()
+                        .setConfigurations(
+                                Collections.singletonList(semanticConfiguration)
+                        );
+
+        SearchIndex newIndex =
+                new SearchIndex(indexName, fields)
+                        .setSemanticSettings(semanticSettings);
+
+        return this._adminClient.createIndex(newIndex);
+    }
+
+    @SuppressWarnings("unchecked")
+    private Mono<String> upsertRecordAsync(String indexName, AzureCognitiveSearchRecord record) {
+        SearchAsyncClient client = this.getSearchClient(indexName);
+        IndexDocumentsOptions throwOnAnyError =
+                new IndexDocumentsOptions().setThrowOnAnyError(true);
+
+        Function<Response<IndexDocumentsResult>, String> transform =
+                (response) -> {
+                    if (response.getStatusCode() == 404
+                            || response.getValue() == null
+                            || response.getValue().getResults() == null
+                            || response.getValue().getResults().isEmpty()) {
+                        throw new AzureCognitiveSearchMemoryException(
+                                "Memory write returned null or an empty set");
+                    }
+                    return response.getValue().getResults().get(0).getKey();
+                };
+
+        return client.mergeOrUploadDocumentsWithResponse(Collections.singleton(record), throwOnAnyError)
+                .map(transform)
+                .onErrorResume(e ->
+                        createIndexAsync(indexName)
+                                .then(client.mergeOrUploadDocumentsWithResponse(
+                                                Collections.singleton(record), throwOnAnyError))
+                                .map(transform)
+                );
+
+    }
+
+    private static MemoryRecordMetadata toMemoryRecordMetadata(AzureCognitiveSearchRecord data) {
+        return new MemoryRecordMetadata(
+                data.isReference(),
+                decodeId(data.getId()),
+                normalizeNullToEmptyString(data.getText()),
+                normalizeNullToEmptyString(data.getDescription()),
+                normalizeNullToEmptyString(data.getExternalSourceName()),
+                normalizeNullToEmptyString(data.getAdditionalMetadata()));
+    }
+
+    private static String normalizeNullToEmptyString(@Nullable String value) {
+        return value != null ? value : EMPTY_STRING;
+    }
+
+    // Normalize index name to match ACS rules.
+    // The method doesn't handle all the error scenarios, leaving it to the service
+    // to throw an error for edge cases not handled locally.
+    // <param name="indexName">Value to normalize</param>
+    // <returns>Normalized name</returns>
+    private static String normalizeIndexName(String indexName) {
+        if (indexName.length() > 128) {
```

Comment:
is 128 a never changing limit?

Reply:
According to https://learn.microsoft.com/en-us/rest/api/searchservice/Naming-rules, index is 2 to 128 characters.